apps/src/lib/util/firehose.js

```diff
@@ -109,14 +109,63 @@ class FirehoseClient {
   }
 
   /**
+   * Returns a unique user ID that is persisted across sessions through local
+   * storage.
+   * WARNING: Mutates local storage if an analytisID has not already been set.
+   * @return {string | null} A unique user ID.
+   */
+  getAnalyticsUuid() {
+    if (!!window.localStorage && !!window.localStorage.getItem("analyticsID")) {
+      return window.localStorage.getItem("analyticsID");
+    }
+    let analytics_uuid = createUuid();
+    if (trySetLocalStorage("analyticsID", analytics_uuid)) {
+      return analytics_uuid;
+    }
+    return null;
+  }
+
+  /**
+   * Returns a hash containing user device information.
+   * storage.
+   * @return {hash} Hash of user device information.
+   */
+  getDeviceInfo() {
+    let device_info = {
+      user_agent: window.navigator.userAgent,
+      window_width: window.innerWidth,
+      window_height: window.innerHeight,
+      hostname: window.location.hostname,
+      full_path: window.location.href
+    };
+    return device_info;
+  }
+
+  /**
+   * Merge various key-value pairs into data.
+   * @param {hash} data The data to add the key-value pairs to.
+   * @return {hash} The data, including the newly added key-value pairs.
+   */
+  addCommonValues(data) {
+    data['created_at'] = new Date().toISOString();
+    data['environment'] = this.getEnvironment();
+    data['uuid'] = this.getAnalyticsUuid();
+    data['device'] = JSON.stringify(this.getDeviceInfo());
```

Comment:
We should verify that this ends up in Redshift in queryable form - I'm not sure if we need to stringify it before sending, we might be double encoding?